c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -471,27 +472,57 @@ impl<'a, 'tcx> ExprRewriteVisitor<'a, 'tcx> {
             self.acx.tcx().erase_regions(to.pointee_ty),
         );
 
+        let orig_from = from;
+        let mut from = orig_from;
+
+        // Ignore differences in lifetimes.
+        if from.pointee_ty != to.pointee_ty {
```

Comment:
check if `erase_regions(from.pointee_ty) == erase_regions(to.pointee_ty)` as well?

Reply:
That's already asserted above - the only difference that's possible at this point is in the lifetimes.  I've updated the comment here to mention this.

Reply:
I think I had the mistaken interpretation that `erase_regions` did erase the lifetimes from the type, if this check compares lifetimes but the assertion above does not. Where can I learn about the difference here? From polonius docs I've gotten used to origins, regions, and lifetimes used interchangeably

Reply:
Not sure what you mean.  `erase_regions(ty)` returns a copy of `ty` where all regions have been erased, e.g. `&'a i32` -> `&'_ i32` (here using `'_` to mean "erased region" rather than "infer this region").

If this function is called with `&'a i32` and `&'b i32`, the assertion above passes because `&'_ i32 == &'_ i32`.  But the check here returns `true` because `&'a i32 != &'b i32` - specifically, `'a != 'b`.  Our goal in this function is to end up with `from == to` (there's a check to this effect at the end), and here we handle the case of differences in the lifetimes, which are irrelevant.

Reply:
I think what was confusing me was that this check was happening at all; if the difference in lifetimes doesn't matter, the assertion is made above that the types are equal, and the only way they can differ here is in the lifetimes, why not unconditionally set `from.pointee_ty = to.pointee_ty`?

Reply:
Ah, good point.  How's this? dc785af4d161e0b3729a42719481850ce9a097f8

Reply:
that lgtm